### PR TITLE
Third and forth Home Work

### PR DESCRIPTION
**Pull request by JavaPsyh**

### CODE REVIEW

**Comment by JavaPsyh** [comment]
Sofia, hi!
Wellcome to my Home Work 3-4. Yes, I put all tasks in one pull request.
Firstly, package name like psyh will be psyh it is very important for me. We can consider it like a domen part of unique class name, but psyh will be psyh. If you want, onetimes I can tell you why this name is that important. Thanks!

**Comment by SevostyanovGerman** [comment]
@JavaPsyh Our task is to get you a job as a developer. Package names are an important part of the agreement between java developers around the world. The correct name is one of the steps to successful employment. I ask you to fulfill the recommendations of Sofia and continue to adhere to our instructions.

File: src/psyh/AsciiCharSequence.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    @Override
+    public char charAt(int index) {
+        return (char) arrayOfBytes[index];
+    }
+
+    @Override
+    public CharSequence subSequence(int start, int end) {
+        byte[] result = new byte[end - start];
```
Comment (Sofasmile) [inline src/psyh/AsciiCharSequence.java]: better add checking:
if (start > end)...

File: src/psyh/CreatSortedArray.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package psyh;
+
+public class CreatSortedArray {
+    public static int[] creatSortedArray(int[] first, int[] second) {
+        if (first == null || second == null) return null;
```
Comment (Sofasmile) [inline src/psyh/CreatSortedArray.java]: cs

File: src/psyh/EqualsAndHachCodeOverriding.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            return true;
+        }
+        if (o == null) {
+            return false;
+        }
+        if (!(o instanceof EqualsAndHachCodeOverriding)) {
+            return false;
+        }
+        EqualsAndHachCodeOverriding otherObj = (EqualsAndHachCodeOverriding) o;
+        if (this.re != otherObj.re) {
```
Comment (Sofasmile) [inline src/psyh/EqualsAndHachCodeOverriding.java]: simplify
if (this.re != otherObj.re || this.im != otherObj.im)...

File: src/psyh/IsStringPalindrome.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package psyh;
+
+public class IsStringPalindrome {
+    public static boolean isStringPalindrome(String text) {
+        if (text == null) return false;
+        String textForAnalysis = text.toLowerCase().replaceAll("[^a-zA-Z0-9]", "");
+        StringBuilder reversOfTextForAnalysis = new StringBuilder();
+        for (int i = textForAnalysis.length() - 1; i >= 0; i--) {
```
Comment (Sofasmile) [inline src/psyh/IsStringPalindrome.java]: optional: you can make it more simple
        String reversOfTextForAnalysis = new StringBuilder(textForAnalysis).reverse().toString();
        return textForAnalysis.equals(reversOfTextForAnalysis);

File: src/psyh/MathSqRt.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package psyh;
+
+public class MathSqRt {
+    public static double mathSqRt(double number) {
+
+        if (number >= 0) {
```
Comment (Sofasmile) [inline src/psyh/MathSqRt.java]: simplify:
if (number < 0) {
            throw new IllegalArgumentException("Expected non-negative number, got " + number);
        } 
return Math.sqrt(number);

File: src/psyh/filterOfComments/KeywordAnalizer.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+abstract class KeywordAnalizer implements TextAnalyzer {
+
+    abstract String[] getKeywords();
+
+    abstract Label getLabel();
+
+    @Override
+    public Label processText(String text) {
+        String[] arrayOfKeywords = getKeywords();
```
Comment (Sofasmile) [inline src/psyh/filterOfComments/KeywordAnalizer.java]: you can use foreach here

File: src/psyh/filterOfComments/Main.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
+package psyh.filterOfComments;
+
+public class Main {
+    public static void main(String[] args) {
+//         инициализация анализаторов для проверки в порядке данного набора анализаторов
```
Comment (Sofasmile) [inline src/psyh/filterOfComments/Main.java]: remove comment or write it in English

File: src/psyh/filterOfComments/Main.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+        for (TextAnalyzer tAnalyzer : textAnalyzer) {
+            Label label = tAnalyzer.processText(text);
+            if (label != Label.OK) return label;
+        }
+        return Label.OK;
+    }
+
+}
+
```
Comment (Sofasmile) [inline src/psyh/filterOfComments/Main.java]: CS

File: src/psyh/fuckingRobot/FuckingRobot.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    public int getX() {
+        return x;
+    }
+
+    public int getY() {
+        return y;
+    }
+
+    public void turnLeft() {
+        System.out.println("Поворот против часовой стрелки");
```
Comment (Sofasmile) [inline src/psyh/fuckingRobot/FuckingRobot.java]: use English

File: src/psyh/fuckingRobot/FuckingRobot.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        if (direction == Direction.DOWN) {
```
Comment (Sofasmile) [inline src/psyh/fuckingRobot/FuckingRobot.java]: better use switch-case

File: src/psyh/fuckingRobot/FuckingRobot.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        if (direction == Direction.RIGHT) {
+            System.out.println("Вправо -> вверх");
+            this.direction = Direction.UP;
+            return;
+        }
+    }
+
+    public void turnRight() {
```
Comment (Sofasmile) [inline src/psyh/fuckingRobot/FuckingRobot.java]: the same

File: src/psyh/fuckingRobot/FuckingRobot.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
+package psyh.fuckingRobot;
+
+import java.util.Random;
+
+public class FuckingRobot {
```
Comment (Sofasmile) [inline src/psyh/fuckingRobot/FuckingRobot.java]: rename your class

File: src/psyh/fuckingRobot/FuckingRobot.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        if (this.direction == Direction.RIGHT) {
+            System.out.println("Вправо -> вниз");
+            this.direction = Direction.DOWN;
+            return;
+        }
+    }
+
+    public void stepForward() {
```
Comment (Sofasmile) [inline src/psyh/fuckingRobot/FuckingRobot.java]: the same

File: src/psyh/fuckingRobot/MoveFuckingRobot.java [hunk 14]
```diff
@@ -0,0 +1,184 @@
+package psyh.fuckingRobot;
+
+public class MoveFuckingRobot {
```
Comment (Sofasmile) [inline src/psyh/fuckingRobot/MoveFuckingRobot.java]: the same

File: src/psyh/fuckingRobot/MoveFuckingRobot.java [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+    private static void wolker(FuckingRobot robot, int distance) {
+        for (; distance > 0; distance--) {
+            robot.stepForward();
+        }
+    }
+
+// Це найгірше, що зі мною траплялося за останній час. Або так код не має писатись, або я не хочу бути програмістом.
+// Софія, пробач, будь ласка, коли я це маніакально писав, а всілякого графоманства було разів у 7 більше,
+// я меньш за все думав, що комусь потім доведеться знов торувати цей нелюдяний шлях.
+    public static void moveFuckingRobot(FuckingRobot robot, int toX, int toY) {
```
Comment (Sofasmile) [inline src/psyh/fuckingRobot/MoveFuckingRobot.java]: this is one of the worst ways to make the robot move
if you don't know how make this method better, I will help you. you can write me.

**Review by Sofasmile — COMMENTED** [review]
resolve comments

File: src/idonotknowhowicannamethispackageoverdotes/robot/MoveRobot.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    private static void wolker(Robot bobot, int distance) {
+        for (; distance > 0; distance--) {
+            bobot.stepForward();
+        }
+    }
+
+    public static void moveRobot(Robot robot, int toX, int toY) {
+        horizontalDistance = robot.getX() > toX ? robot.getX() - toX : toX - robot.getX();
```
Comment (Sofasmile) [inline src/idonotknowhowicannamethispackageoverdotes/robot/MoveRobot.java]: better write 
horizontalDistance = Math.abs (toX - robot.getX());

File: src/idonotknowhowicannamethispackageoverdotes/robot/MoveRobot.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+        verticalDistance = robot.getY() > toY ? robot.getY() - toY : toY - robot.getY();
+        Direction longWay = horizontalDistance > verticalDistance ?
```
Comment (Sofasmile) [inline src/idonotknowhowicannamethispackageoverdotes/robot/MoveRobot.java]: simplify, make it more understandable (without ternary operator)

Comment (Sofasmile) [reply src/idonotknowhowicannamethispackageoverdotes/robot/MoveRobot.java]: why you use longWay and shortWay?

**Review by Sofasmile — COMMENTED** [review]
resolve comments

File: src/tasks/AsciiCharSequence.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
+package tasks;
+
```
Comment (JavaPsyh) [inline src/tasks/AsciiCharSequence.java]: Sorry!

**Comment by Sofasmile** [comment]
why you did not correct comments?

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**